File: Examples/Mvc.VB/Controllers/DocumentConverterController.Overview.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.IO
Imports System.Web
Imports GleamTech.DocumentUltimate
Imports GleamTech.DocumentUltimateExamples.Mvc.VB.Models
Imports GleamTech.ExamplesCore
Imports GleamTech.IO
Imports GleamTech.Util
Imports GleamTech.Web

Namespace Controllers
    Partial Public Class  DocumentConverterController
        Inherits Controller
	    Public Function Overview() As ActionResult
		    Dim model = New OverviewViewModel() With {
			    .ExampleFileSelector = New ExampleFileSelector() With {
				    .ID = "exampleFileSelector",
				    .InitialFile = "Default.pdf"
			    }
		    }

		    Dim inputDocument = model.ExampleFileSelector.SelectedFile
		    Dim fileInfo = New FileInfo(inputDocument)
		    Dim inputFormat = DocumentFormatInfo.[Get](inputDocument)
		    model.InputFormat = If(inputFormat IsNot Nothing, inputFormat.Description, "(not supported)")

		    PopulatePossibleOutputFormats(inputDocument, model)

		    model.ConvertHandlerUrl = ExamplesCoreConfiguration.GetDynamicDownloadUrl(ConvertHandlerName, 
                New NameValueCollection() From {
			        {"inputDocument", ExamplesCoreConfiguration.ProtectString(inputDocument)},
			        {"version", fileInfo.LastWriteTimeUtc.Ticks & "-" & fileInfo.Length}
		        })

		    Return View(model)
	    End Function

	    Private Sub PopulatePossibleOutputFormats(inputDocument As String, model As OverviewViewModel)
		    For Each format As DocumentFormat In DocumentConverter.EnumeratePossibleOutputFormats(inputDocument)
			    Dim formatInfo = DocumentFormatInfo.[Get](format)

			    Dim groupData As List(Of SelectListItem) = Nothing
			    If Not model.OutputFormats.TryGetValue(formatInfo.Group.Description, groupData) Then
				    groupData = New List(Of SelectListItem)()
				    model.OutputFormats.Add(formatInfo.Group.Description, groupData)
			    End If
			    groupData.Add(New SelectListItem() With {
				    .Text = formatInfo.Description,
				    .Value = formatInfo.Value.ToString()
			    })
		    Next

		    If model.OutputFormats.Count = 0 Then
			    model.OutputFormats.Add("(not supported)", New List(Of SelectListItem)())
		    End If
	    End Sub

	    Public Shared Sub ConvertHandler(context As HttpContext)
		    Dim result As DocumentConverterResult

		    Try
			    Dim inputDocument = New BackSlashPath(ExamplesCoreConfiguration.UnprotectString(context.Request("inputDocument")))
			    Dim outputFormat = DirectCast([Enum].Parse(GetType(DocumentFormat), context.Request("outputFormat")), DocumentFormat)
			    Dim fileName = inputDocument.FileNameWithoutExtension + "." + DocumentFormatInfo.[Get](outputFormat).DefaultExtension
			    Dim outputPath = ConvertedPath.Append(context.Session.SessionID).Append(fileName)
			    Dim outputDocument = outputPath.Append(fileName)

			    If Directory.Exists(outputPath) Then
				    Directory.Delete(outputPath, True)
			    End If
			    Directory.CreateDirectory(outputPath)
			    result = DocumentConverter.Convert(inputDocument, outputDocument, outputFormat)
		    Catch exception As Exception
			    context.Response.Write("<span style=""color: red; font-weight: bold"">Conversion failed</span><br/>")
			    context.Response.Write(exception.Message)
			    Return
		    End Try

		    context.Response.Write("<span style=""color: green; font-weight: bold"">Conversion successful</span>")
		    context.Response.Write("<br/>Conversion time: " & result.ElapsedTime.ToString())
		    context.Response.Write("<br/>Output files:")

		    If result.OutputFiles.Length > 1 Then
			    context.Response.Write(Convert.ToString(" - ") & GetZipDownloadLink(New FileInfo(result.OutputFiles(0)).Directory))
		    End If

		    context.Response.Write("<br/><ol>")
		    For Each outputFile As String In result.OutputFiles
			    If outputFile.EndsWith("\") Then
				    Dim directoryInfo = New DirectoryInfo(outputFile)
				    context.Response.Write(String.Format("<br/><li><b>{0}\</b> - {1}</li>", directoryInfo.Name, GetZipDownloadLink(directoryInfo)))
			    Else
				    Dim fileInfo = New FileInfo(outputFile)
				    context.Response.Write(String.Format("<br/><li><b>{0}</b> ({1} bytes) - {2}</li>", fileInfo.Name, fileInfo.Length, GetDownloadLink(fileInfo)))
			    End If
		    Next
		    context.Response.Write("<br/></ol>")
	    End Sub

	    Private Shared Function GetDownloadLink(fileInfo As FileInfo) As String
		    Return String.Format("<a href=""{0}"">Download</a>", ExamplesCoreConfiguration.GetDownloadUrl(fileInfo.FullName, fileInfo.LastWriteTimeUtc.Ticks.ToString()))
	    End Function

	    Private Shared Function GetZipDownloadLink(directoryInfo As DirectoryInfo) As String
		    Return String.Format("<a href=""{0}"">Download as Zip</a>", ExamplesCoreConfiguration.GetDynamicDownloadUrl(ZipDownloadHandlerName, New NameValueCollection() From {
			    {"path", ExamplesCoreConfiguration.ProtectString(directoryInfo.FullName)},
			    {"version", directoryInfo.LastWriteTimeUtc.Ticks.ToString()}
		    }))
	    End Function

	    Public Shared Sub ZipDownloadHandler(context As HttpContext)
		    Dim path = New BackSlashPath(ExamplesCoreConfiguration.UnprotectString(context.Request("path"))).RemoveTrailingSlash()

		    Dim fileResponse = New FileResponse(context, 0)
		    fileResponse.Transmit(Sub(targetStream, copyFileCallback) 
		        QuickZip.Zip(targetStream, Directory.EnumerateFileSystemEntries(path))
            End Sub, path.FileName + ".zip", 0)

	    End Sub

	    Private Shared ReadOnly Property ConvertHandlerName() As String
		    Get
			    If m_convertHandlerName Is Nothing Then
				    m_convertHandlerName = "ConvertHandler"
				    ExamplesCoreConfiguration.RegisterDynamicDownloadHandler(m_convertHandlerName, AddressOf ConvertHandler)
			    End If

			    Return m_convertHandlerName
		    End Get
	    End Property
	    Private Shared m_convertHandlerName As String

	    Private Shared ReadOnly Property ZipDownloadHandlerName() As String
		    Get
			    If m_zipDownloadHandlerName Is Nothing Then
				    m_zipDownloadHandlerName = "ZipDownloadHandler"
				    ExamplesCoreConfiguration.RegisterDynamicDownloadHandler(m_zipDownloadHandlerName, AddressOf ZipDownloadHandler)
			    End If

			    Return m_zipDownloadHandlerName
		    End Get
	    End Property
	    Private Shared m_zipDownloadHandlerName As String

	    Private Shared ReadOnly ConvertedPath As BackSlashPath = HostingPathHelper.MapPath("~/App_Data/ConvertedDocuments")
    End Class
End Namespace

File: Examples/AspNetCoreCS/Controllers/DocumentConverterController.Overview.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using GleamTech.AspNet;
using GleamTech.DocumentUltimate;
using GleamTech.DocumentUltimateExamples.AspNetCoreCS.Models;
using GleamTech.Examples;
using GleamTech.IO;
using GleamTech.Zip;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.Rendering;

namespace GleamTech.DocumentUltimateExamples.AspNetCoreCS.Controllers
{
    public partial class DocumentConverterController
    {
        public IActionResult Overview()
        {
            var model = new OverviewViewModel
            {
                ExampleFileSelector = new ExampleFileSelector
                {
                    Id = "exampleFileSelector",
                    InitialFile = "Default.pdf"
                }
            };

            var inputDocument = model.ExampleFileSelector.SelectedFile;
            var fileInfo = new FileInfo(inputDocument);
            var inputFormat = DocumentFormatInfo.Get(inputDocument);
            model.InputFormat = inputFormat != null ? inputFormat.Description : "(not supported)";

            PopulatePossibleOutputFormats(inputDocument, model);

            model.ConvertHandlerUrl = ExamplesConfiguration.GetDynamicDownloadUrl(
                ConvertHandlerName,
                new Dictionary<string, string>
                {
                    {"inputDocument", ExamplesConfiguration.ProtectString(inputDocument)},
                    {"version", fileInfo.LastWriteTimeUtc.Ticks + "-" +  fileInfo.Length}
                });

            return View(model);
        }

        private void PopulatePossibleOutputFormats(string inputDocument, OverviewViewModel model)
        {
            foreach (var format in DocumentConverter.EnumeratePossibleOutputFormats(inputDocument))
            {
                var formatInfo = DocumentFormatInfo.Get(format);

                List<SelectListItem> groupData;
                if (!model.OutputFormats.TryGetValue(formatInfo.Group.Description, out groupData))
                {
                    groupData = new List<SelectListItem>();
                    model.OutputFormats.Add(formatInfo.Group.Description, groupData);
                }
                groupData.Add(new SelectListItem
                {
                    Text = formatInfo.Description,
                    Value = formatInfo.Value.ToString()
                });
            }

            if (model.OutputFormats.Count == 0)
                model.OutputFormats.Add("(not supported)", new List<SelectListItem>());
        }

        public static void ConvertHandler(IHttpContext context)
        {
            DocumentConverterResult result;

            var inputDocument = new PhysicalPath(ExamplesConfiguration.UnprotectString(context.Request["inputDocument"]));
            var outputFormat = (DocumentFormat)Enum.Parse(typeof(DocumentFormat), context.Request["outputFormat"]);
            var fileName = inputDocument.FileNameWithoutExtension + "." + DocumentFormatInfo.Get(outputFormat).DefaultExtension;
            var outputPath = ConvertedPath.Append(context.Session.Id).Append(fileName);
            var outputDocument = outputPath.Append(fileName);

            try
            {
                if (Directory.Exists(outputPath))
                    Directory.Delete(outputPath, true);
                Directory.CreateDirectory(outputPath);

                result = DocumentUltimate.DocumentConverter.Convert(inputDocument.ToString(), outputDocument.ToString(), outputFormat);
            }
            catch (Exception exception)
            {
                context.Response.Output.Write("<span style=\"color: red; font-weight: bold\">Conversion failed</span><br/>");
                context.Response.Output.Write(exception.Message);
                return;
            }

            context.Response.Output.Write("<span style=\"color: green; font-weight: bold\">Conversion successful</span>");
            context.Response.Output.Write("<br/>Conversion time: {0}", result.ElapsedTime);
            context.Response.Output.Write("<br/>Output files:");

            if (result.OutputFiles.Length > 1)
                context.Response.Output.Write(" - " + GetZipDownloadLink(new DirectoryInfo(outputPath)));

            context.Response.Output.Write("<br/><ol>");
            foreach (var outputFile in result.OutputFiles)
            {
                var outputFilePath = outputPath.Append(outputFile).ToString();

                if (outputFilePath.EndsWith("\\"))
                {
                    var directoryInfo = new DirectoryInfo(outputFilePath);
                    context.Response.Output.Write(
                        "<br/><li><b>{0}\\</b> - {1}</li>",
                        directoryInfo.Name,
                        GetZipDownloadLink(directoryInfo));
                }
                else
                {
                    var fileInfo = new FileInfo(outputFilePath);
                    context.Response.Output.Write(
                        "<br/><li><b>{0}</b> ({1} bytes) - {2}</li>",
                        fileInfo.Name,
                        fileInfo.Length,
                        GetDownloadLink(fileInfo));
                }
            }
            context.Response.Output.Write("<br/></ol>");
        }

        private static string GetDownloadLink(FileInfo fileInfo)
        {
            return string.Format(
                "<a href=\"{0}\">Download</a>",
                ExamplesConfiguration.GetDownloadUrl(fileInfo.FullName, fileInfo.LastWriteTimeUtc.Ticks.ToString()));
        }

        private static string GetZipDownloadLink(DirectoryInfo directoryInfo)
        {
            return string.Format(
                "<a href=\"{0}\">Download as Zip</a>",
                ExamplesConfiguration.GetDynamicDownloadUrl(
                    ZipDownloadHandlerName,
                    new Dictionary<string, string>
                    {
                        {"path", ExamplesConfiguration.ProtectString(directoryInfo.FullName)},
                        {"version", directoryInfo.LastWriteTimeUtc.Ticks.ToString()},
                    }));
        }

        public static void ZipDownloadHandler(IHttpContext context)
        {
            var path = new PhysicalPath(ExamplesConfiguration.UnprotectString(context.Request["path"])).RemoveTrailingSlash();

            var zipFile = path.Append(path.FileName + ".zip");
            var itemPaths = Directory.EnumerateFileSystemEntries(path)
                .Where(p => p != zipFile);

            QuickZip.Zip(zipFile, itemPaths);

            var fileResponse = new FileResponse(context);
            fileResponse.Transmit(zipFile);
        }

        private static string ConvertHandlerName
        {
            get
            {
                if (convertHandlerName == null)
                {
                    convertHandlerName = "ConvertHandler";
                    ExamplesConfiguration.RegisterDynamicDownloadHandler(convertHandlerName, ConvertHandler);
                }

                return convertHandlerName;
            }
        }
        private static string convertHandlerName;

        private static string ZipDownloadHandlerName
        {
            get
            {
                if (zipDownloadHandlerName == null)
                {
                    zipDownloadHandlerName = "ZipDownloadHandler";
                    ExamplesConfiguration.RegisterDynamicDownloadHandler(zipDownloadHandlerName, ZipDownloadHandler);
                }

                return zipDownloadHandlerName;
            }
        }
        private static string zipDownloadHandlerName;

        private static readonly PhysicalPath ConvertedPath = Hosting.ResolvePhysicalPath("~/App_Data/ConvertedDocuments");
    }
}
